File: built-in plugins dsp/mb-ethnic-balalaika.cpp
/**
 * MB Balalaika
 * Category : instrument
 * Type     : ethnic
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Russian triangular balalaika with bright strumming
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ETHNIC_BALALAIKA_H
#define MB_ETHNIC_BALALAIKA_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEthnicBalalaika : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-ethnic-balalaika";
    static constexpr const char* PLUGIN_NAME    = "MB Balalaika";
    static constexpr const char* PLUGIN_TYPE    = "ethnic";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.7f;  // range [0, 1]
    float tremolo = 0.5f;  // range [0, 1]
    float body = 0.4f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbEthnicBalalaika() = default;
    ~MbEthnicBalalaika() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.tremolo = std::clamp(params.tremolo, 0f, 1f);
        params.body = std::clamp(params.body, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Balalaika
        return input;
    }
};

#endif // MB_ETHNIC_BALALAIKA_H

File: built-in plugins dsp/mb-ethnic-balalaika.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbEthnicBalalaikaPlugin: PluginDefinition = {
  id: "mb-ethnic-balalaika",
  slug: "mb-ethnic-balalaika",
  name: "MB Balalaika",
  category: "instrument",
  type: "ethnic" as any,
  version: "1.0.0",
  description: "Russian triangular balalaika with bright strumming",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "triangle", detune: -3, gain: 0.4 },
    { type: "triangle", detune: 3, gain: 0.4 },
    { type: "sine", detune: 1200, gain: 0.2 },
  ],
  envelope: { attack: 0.001, decay: 0.35, sustain: 0.15, release: 0.25 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tremolo",
      name: "Tremolo",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "body",
      name: "Body",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { brightness: 0.7, tremolo: 0.5, body: 0.4, volume: 0.8 },
};

export default MbEthnicBalalaikaPlugin;
